### PR TITLE
[DDO-2166] Sign and notarize macOS releases + extras!

### PR DESCRIPTION
**Pull request by em-may**
This PR adds signing and notarizing for macOS builds. It also adds the following goodies:
* Go deps are now cached between builds
* Builds are parellelized
* The docker build uses the linux release artifact instead of doing another build

Overall this speeds up builds a lot and minimizes the time impact that adding signing/notarizing would otherwise have.

Some notes on time improvements/speed ups, using [this](https://github.com/broadinstitute/thelma/runs/7362013023?check_suite_focus=true) before example and [this](https://github.com/broadinstitute/thelma/actions/runs/2707986760) after example:
* Build times
  * Before: 5m 53s
  * **After: 1m59s**
* Signing and notarizing:
  * Before: N/A
  * After: 4m 35s
* Image gen + push + release push:
  * Before: 3m 43s
  * **After: 1m 47s**
* Total build GHA run times:
  * Before: 10m 0s
  * After: 10m 0s

TL;DR: Signing and notarizing for macOS releases now works and it adds basically no time to the build workflow despite adding a step that takes 4.5mins.

### CODE REVIEW

File: .github/actions/make/action.yaml
Comment (choover-broad) [inline .github/actions/make/action.yaml:30]: do we need a `cache: true` here to actually cache dependencies?

File: .github/workflows/build.yaml
Comment (choover-broad) [inline .github/workflows/build.yaml:55]: so just checking my understanding, this isn't actually caching as much as "save contents of build output directory and make available for the notarization job"?

File: .github/workflows/build.yaml
Comment (choover-broad) [inline .github/workflows/build.yaml:117]: maybe call the two "caches" `unsigned-build-artifacts` and `signed-build-artifacts` for clarity? IDK, it's a matter of taste but that would have helped me personally understand sooner 🤷  (Since caching is not what we're doing as much as copying artifacts across jobs)

File: Makefile
Comment (choover-broad) [inline Makefile:42]: One thing that's confusing to me about these changes is that even though the Makefile has a MACOS_SIGN_AND_NOTARIZE variable, if I set it to true, my binaries are not going to be notarized :) All of that happens elsewhere, in GHA pipelines, right?

I almost think it might be more clear to not change this file... We can leave it as-is, right? It's not like producing the tarball deletes the unsigned binaries from the cache...

File: Dockerfile
Comment (choover-broad) [inline Dockerfile:10]: Just want to say that this alone is a fantastic speedup! So great to see the binaries re-used instead of another compile! Thank you!

File: .github/workflows/build.yaml
Comment (choover-broad) [inline .github/workflows/build.yaml:144]: ~~I have a very stupid question. Does output here contain the `arm64` or `amd64` thelma? Is it just whatever was most recently compiled? I don't understand how `sign-and-notarize` could be run in succession on the same output directory and produce results for different platforms :confused:~~

File: Makefile
Comment (choover-broad) [reply Makefile:42]: OK actually I think this variable might never be set to true, since it looks like `sign-and-notarize.sh` unpacks and repacks the tarball. So it seems like we could definitely leave as-is.

File: .github/workflows/build.yaml
Comment (choover-broad) [reply .github/workflows/build.yaml:144]: Okay I was confused, it looks like the ./output parameter here is just a tmp directory and not input. The 2nd param is the name of the tarball to replace (so input and output)

**Review by choover-broad — APPROVED** [review]
Looks good! Great to see such a huge bonus perf improvement in thelma's build process!

File: .github/actions/make/action.yaml
Comment (em-may) [reply .github/actions/make/action.yaml:34]: Nope! Turns out if you do that it will cache between workflows, which we don't want. By default, it uses `actions/cache` (I think) under the hood to give the behavior we want (i.e. caching between jobs within the workflow). The proof is that the build jobs no longer have output showing that they are pulling Go deps.

File: .github/workflows/build.yaml
Comment (em-may) [reply .github/workflows/build.yaml:270]: In the time since this comment, all of the intermediary caches have been given platform-specific names and only the final release cache is called `release-cache`. I've also added a big comment to the top of this file explaining how the caches work for future reference. I think that should do it.

File: scripts/sign-and-notarize.sh
Comment (em-may) [reply scripts/sign-and-notarize.sh:82]: Apple notarizes all the files that we upload in the zip, and since notarization requires uploading a file to Apple and waiting for their approval, I think it's best to do it all at once. TL;DR: No, we do not need to notarize each file individually.

File: scripts/sign-and-notarize.sh
Comment (em-may) [reply scripts/sign-and-notarize.sh:105]: If you look further down you'll see some logic for determining if the response was good or bad. If the submission succeeded, then it's not printed, if it failed, it prints. Initially, I mostly did this to make the output cleaner, and it works, so unless there's a strong desire to change it I'm inclined to leave it since it works.

File: .github/workflows/build.yaml
Comment (em-may) [reply .github/workflows/build.yaml]: Correct, the first param is the output dir, so that a working dir for the script can be created inside it - the second and third are input and output tarballs.

**Pull request merged**
